archinfo/arch_arm.py

```diff
@@ -274,7 +281,83 @@ class ArchARMEL(ArchARM):
     ld_linux_name = 'ld-linux.so.3'
     elf_tls = TLSArchInfo(1, 8, [], [0], [], 0, 0)
 
+class ArchARMCortexM(ArchARMEL):
+    """
+    This is an architecture description for ARM Cortex-M microcontroller-class CPUs.
+
+    These CPUs have the following unusual / annoying distinctions from their relatives:
+    - Explicitly only support the Thumb-2 instruction set.  Executing with the T-bit off causes the processor to fault
+    instantly
+    - Always little-endian
+    - Coprocessors? Nope, none of that rubbish
+    - Well-known standard memory map across all devices
+    - Rarely use an MPU, even though this does exist on some devices
+    - A built-in "NVIC" (Nested Vectored Interrupt Controller) as part of the standard.
+    - Standardized "blob format" including the IVT, with initial SP and entry prepended
+    - Usually don't run an OS (SimLinux? No thanks)
+    - As part of the above, handle syscalls (SVC) instructions through an interrupt (now called PendSV)
+    Uses its own fancy stack layout for this, which (UGH) varies by sub-sub-architecture
+    - Some fancy instructions normally never seen in other uses of Thumb (CPSID, CPSIE, WFI, MRS.W, MSR.W)
+    - New registers, namely:
+    * FAULTMASK
+    * PRIMASK
+    * BASEPRI
+    * CONTROL
+    * SP, banked as PSP or MSP
+    * PSR, now just one PSR, with a few meta-registers APSR, IPSR, and EPSR which take a chunk of that each
+    
+    """
+
+    name = "ARMCortexM"
+    triplet = 'arm-none-eabi'  # ARM's own CM compilers use this triplet
+
+    # These are the standard THUMB prologs.  We leave these off for other ARMs due to their length
+    # For CM, we assume the FPs are OK, as they are almost guaranteed to appear all over the place
+    function_prologs = {}
+
+    thumb_prologs = {
+        br"[\x00-\xff]\xb5",  # push {xxx,lr}
+        br"[\x00-\xff][\x00-\xff]\x2d\xe9", # push.w {xxx, lr}
+    }
+    function_epilogs = {
+        br"[\x00-\xff]\xbd" # pop {xxx, pc}
+        # TODO: POP.W
+    }
+
+    def __init__(self, *args, **kwargs):
+        ArchARMEL.__init__(self, *args, **kwargs)
+
+        # See http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/CHDBIBGJ.html
+        # for relevant pain-points.  Most of these only get updated on an exception/interrupt/syscall
+        # Except we leave T on because we have to.
+        self.register_list += [
```

Comment:
Shouldn't this happen before the super-call to `__init__`? the base initializer does a lot of processing with the registers, and it should be where e.g. the register file allocation for non-vex registers happens (where are these assigned offsets, anyway..?)

Reply:
Done